src/durable_write.c

```diff
@@ -0,0 +1,993 @@
+        }
+    }
+    return 0;
+}
+
+/*================================= Replica offset management =============== */
+
+/*
+ * Populates the offset of each replica. If the replica is offline,
+ * then the function places a ZERO for its entry.
+ * @offsets        The array that needs to be filled in. Function assumes that proper memory has been allocated for it.
+ * @numReplicas    The size of the offsets array that needs to be filled in. 
+ */
+static void populateReplicaOffsets(long long *offsets, const size_t numReplicas) {
+    memset(offsets, 0, sizeof(long long) * numReplicas);
+ 
+    // iterate the replicas to get the offset they have reached.
+    listIter li;
+    listRewind(server.replicas, &li);
+ 
+    listNode *ln = listNext(&li);
+    for (unsigned i=0; i < numReplicas && ln != NULL; ln = listNext(&li), i++) {
+        const client *replica = listNodeValue(ln);
+        serverAssert(replica->repl_data);
+        if (replica->repl_data->repl_state == REPLICA_STATE_ONLINE) {
+            offsets[i] = replica->repl_data->repl_ack_off;
+        }
+    }
+}
+
+/**
+ * This function loops through all the replica's offset, finding the max offset that the required replicas have acknowledged.
+ * In case the required replicas exceeds the number of replicas connected, the function will return 0 indicating the offset
+ * is not reached by sufficient amount of replicas.
+ * 
+ * @param numAcksNeeded The number of replicas that need to acknowledge the offset.
+ * @returns The offset that requested replicas have reached. 
+ *          In absence of required replicas, the primary offset is returned.
+ *          If there is not enough number of replicas connected, return -1.
+ */ 
+long long getConsensusOffset(const unsigned long numAcksNeeded) {
+    const unsigned long numReplicas = listLength(server.replicas);
+    if (numAcksNeeded == 0) {
+        // If no ack is needed, then the consensus offset is the one primary is at.
+        return server.primary_repl_offset;
+    }
+
+    // If the number of connected replicas is less than the number of required replicas,
+    // return -1 because we don't have enough number of replicas for the ACK. 
+    if (numReplicas < numAcksNeeded) {
+        return -1;
+    }
+
+    // fetch and sort the replica offsets for all replicas. This way we get the top offsets that have been acknowledged.
+    // Kth element in the sorted array will give us the offset that has been acknowledged by K replicas.
+    durable_t *durability = &server.durability;
+
+    // make sure we have enough space for the replicas. Resize only if the required 
+    // replica count is larger. No need to downsize.
+    if(durability->replica_offsets_size < numReplicas) {
+        durability->replica_offsets = zrealloc(durability->replica_offsets, numReplicas * sizeof(long long));
+        durability->replica_offsets_size = numReplicas;
+    }
+    populateReplicaOffsets(durability->replica_offsets, numReplicas);
```

Comment:
It looks like `durability->replica_offsets` is only used within this function. Instead of a global and heap-allocated, we can turn this into a variable-length array on the stack.

```suggestion
    long long replica_offsets[numReplicas];
    populateReplicaOffsets(replica_offsets, numReplicas);
```